intake_xarray/image.py

```diff
@@ -0,0 +1,265 @@
+            - ``s3://data/*.jpeg``
+            - ``https://example.com/image.png`
+            - ``s3://data/Images/{{ landuse }}/{{ landuse }}{{ '%02d' % id }}.tif``
+    chunks: int or dict
+        Chunks is used to load the new dataset into dask
+        arrays. ``chunks={}`` loads the dataset with dask using a single
+        chunk for all arrays.
+    path_as_pattern: bool or str, optional
+        Whether to treat the path as a pattern (ie. ``data_{field}.tif``)
+        and create new coodinates in the output corresponding to pattern
+        fields. If str, is treated as pattern to match on. Default is True.
+    concat_dim : str or iterable
+        Dimension over which to concatenate. If iterable, all fields must be
+        part of the the pattern.
+    xarray_kwargs : dict, optional
+        Any further arguments to pass to reader function.
+    """
+    name = 'xarray_image'
+
+    def __init__(self, urlpath, chunks=None, concat_dim='concat_dim',
+                 xarray_kwargs=None, metadata=None, path_as_pattern=True,
+                 storage_options=None, **kwargs):
+        self.path_as_pattern = path_as_pattern
+        self.urlpath = urlpath
+        self.chunks = chunks
+        self.concat_dim = concat_dim
+        self.storage_options = storage_options or {}
+        self._kwargs = xarray_kwargs or kwargs
+        self._ds = None
+        super(ImageSource, self).__init__(metadata=metadata)
+
+    def _open_files(self, files):
+        import pandas as pd
+        from xarray import DataArray
+
+        out = multireader(files, self.chunks, self.concat_dim, **self._kwargs)
+        if not self.pattern:
+            return out
+
+        coords = {}
+        filenames = [f.path for f in files]
+        field_values = reverse_formats(self.pattern, filenames)
+
+        if isinstance(self.concat_dim, list):
+            if not set(field_values.keys()).issuperset(set(self.concat_dim)):
+                raise KeyError('All concat_dims should be in pattern.')
+            index = pd.MultiIndex.from_tuples(
+                zip(*(field_values[dim] for dim in self.concat_dim)),
+                names=self.concat_dim)
+            coords = {
+                k: DataArray(v, dims=('dim_0'))
+                for k, v in field_values.items() if k not in self.concat_dim
+            }
+            out = (out.assign_coords(dim_0=index, **coords)  # use the index
+                      .unstack().chunk(self.chunks))  # unstack along new index
+            return out.transpose(*self.concat_dim,  # reorder dims
+                                 *filter(lambda x: x not in self.concat_dim,
+                                         out.dims))
+        else:
+            coords = {
+                k: DataArray(v, dims=self.concat_dim)
+                for k, v in field_values.items()
+            }
+            return out.assign_coords(**coords).chunk(self.chunks)
+
+    def _open_dataset(self):
+        from dask.bytes import open_files
```

Comment:
This is the main entry function the find a set of files and pass them to the reader ...